fix(file): incorrectly detects file changes on Windows when isExecutable is set

gmeligio opened the pull request:
## Motivation

The `package.json` file is marked as executable in a `NodeProject` when running on Windows, but the synthesized one is not executable. Then, this mismatch causes the package manager to rerun every time `projen` runs with unnecessary additional time. It detects correctly if the file is executable on Linux and the issue is only for Windows.

The function that checks is
https://github.com/projen/projen/blob/f8a1ac04d4a635f39a511f2af205dc0a75294d74/src/util.ts#L421-L428

The cause is that it's the same as checking for file existence on Windows, as mentioned in https://nodejs.org/api/fs.html#file-access-constants
>  X_OK: ... . This has no effect on Windows (will behave like fs.constants.F_OK).

The way to correctly detect if a file is executable in Windows is by checking the file extension in the environment variable `PATHEXT`, as [defined in PowerShell help](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_environment_variables?view=powershell-7.3#path-information).

More generally, this means that all files that are checked by projen for the attribute executable will be marked as such on Windows if they exist when it's not the case.

## Changes

1. Add function `assertExecutablePermissions`.
1. Override File.executable with `assertExecutablePermissions` in File's constructor.
1. Override isExecutable with `assertExecutablePermissions` in util.isExecutable.
1. Unrelated small change: removed a console.log in `test/javascript/eslint.test.ts` that I added accidentally in another PR and got merged.

## Test

1. After the changes in the `projen` repository in WSL locally, run `yarn package:js` to create projen@0.0.0.jsii.tgz
1. Copy projen@0.0.0.jsii.tgz from WSL to Windows.
1. In `my-node-project` in Windows run `yarn install projen@0.0.0.jsii.tgz`
1. Set logging to DEBUG

	```ts
	logging: {
	    level: LogLevel.DEBUG,
	  },
	```

1. Run `yarn projen` several times to check the result.
1. The package manager doesn't run because package.json hasn't changed.
1. Also, the terminal output confirms that package.json is detected as non-executable:
```bash
👾 no change in D:\my-projen-project\package.json
```

---
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

mrgrain commented:
This is awesome, thank you! I'll need a bit of time to review it.

I do wonder if we just want to bypass the check on windows though (maybe warn?)?
Because in the current implementation, you can set any file to being executable - however on windows it would need to be the correct file ending. So `new File('foobar.js', { executable: true })` would have no effect and probably still cause the change condition to trigger.

gmeligio commented:
Thank you for the explanation. Bypassing the `executable` check will work indeed. I guess that change will be somewhere around here 
https://github.com/projen/projen/blob/34edf99eaa7884cc279e946c56da8e322e6dc839/src/file.ts#L158-L171

Because Windows doesn't save the attribute on the file, but on the current installation, another option could be to use `git` to keep the file attributes. This variant will be cross-platform but I haven't used it before. Please take a look at the image below.

<img width="437" alt="image" src="https://github.com/projen/projen/assets/22875166/7bdb0b68-63be-40de-964f-7944588da407">

Indeed, bypassing the executable attribute is easier than refactoring the attribute checking to git. Please let me know what you think, and I can turn the PR in the bypass direction if you agree.

mrgrain commented:
Apologies for the late response, I was on annual leave. Let's bypass this by introducing a new helper function:
```ts
function assertExecutablePermissions(filePath: string, shouldBeExecutable: boolean): boolean; 
```
that clearly states in it's doc string that it always returns true on Windows.

gmeligio commented:
No worries. Sounds great! Thank you for the idea. I'll do that during this or next week.

gmeligio commented:
I added an implementation to bypass. Please, @mrgrain, take a look when you have time to see if it matches your proposal.

I modified your proposed signature
```ts
function assertExecutablePermissions(filePath: string, shouldBeExecutable: boolean): boolean; 
```
with 
```ts
function assertExecutablePermissions(shouldBeExecutable: boolean): boolean; 
```
Because when implementing, I didn't see the need for `filePath`. If it's in a `new File('foobar.js', { executable: true })`, the path may not exist, and I think it should not be checked. If it's in `isExecutable`, the path was already checked in that function and doesn't need to be passed to `assertExecutablePermissions`. So, I removed the `filePath` parameter at the end.

Please let me know if my interpretation needs to be corrected. I appreciate it very much.

mrgrain reviewed (changes requested):
Made an inline comment, basically I'd like us to keep `this.executable` as is and only change the check inside `synthesize`. On Unix system, the check should stay as it is. On Windows we cannot reliably check, so the check needs to always pass.

gmeligio commented on src/file.ts:
```diff
@@ -108,7 +114,7 @@ export abstract class FileBase extends Component {
 
     this.readonly = !project.ejected && (options.readonly ?? true);
-    this.executable = options.executable ?? false;
+    this.executable = assertExecutablePermissions(options.executable);
```
Makes sense. Initially, I was overriding the user intention to make sure the state had also the bypass. You're right this would be a breaking change for existing projects that use `file.executable`.

I reverted back to the previous code.

gmeligio commented:
Thank you for the review. I fixed the bypass to be only on `synthesize`. Now I see why the `filePath` parameter is needed in the `assertExecutablePermissions` function.

mrgrain reviewed (approved):
Perfect thank you!
